OpenRA.Mods.RA2/Traits/BaseSpawnerParent.cs

```diff
@@ -0,0 +1,271 @@
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.RA2.Traits
+{
+	// What to do when parent is killed or mind controlled
+	public enum SpawnerChildDisposal
+	{
+		DoNothing,
+		KillChildren,
+		GiveChildrenToAttacker
+	}
+
+	public class BaseSpawnerChildEntry
+	{
+		public string ActorName = null;
+		public Actor Actor = null;
+		public BaseSpawnerChild SpawnerChild = null;
+
+		public bool IsValid { get { return Actor != null && !Actor.IsDead; } }
+	}
+
+	[Desc("This actor can spawn actors.")]
+	public class BaseSpawnerParentInfo : ConditionalTraitInfo
+	{
+		[ActorReference]
+		[FieldLoader.Require]
+		[Desc("Spawn these units.")]
+		public readonly string[] Actors;
+
+		[Desc("Child actors to contain upon creation. Set to -1 to start with full children.")]
+		public readonly int InitialActorCount = -1;
+
+		[Desc("Name of the armaments that grant this condition.")]
+		public readonly HashSet<string> ArmamentNames = new HashSet<string>() { "primary" };
+
+		[Desc("What happens to the children when the parent is killed?")]
+		public readonly SpawnerChildDisposal ChildDisposalOnKill = SpawnerChildDisposal.KillChildren;
+
+		[Desc("What happens to the children when the parent is mind controlled?")]
+		public readonly SpawnerChildDisposal ChildDisposalOnOwnerChange = SpawnerChildDisposal.GiveChildrenToAttacker;
+
+		[Desc("Only spawn initial load of children?")]
+		public readonly bool NoRegeneration = false;
+
+		[Desc("Spawn all children at once when regenerating children, instead of one by one?")]
+		public readonly bool SpawnAllAtOnce = false;
+
+		[Desc("Spawn regen delay, in ticks")]
+		public readonly int RespawnTicks = 150;
+
+		public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			base.RulesetLoaded(rules, ai);
+
+			if (InitialActorCount > Actors.Length)
+				throw new YamlException("InitialActorCount can't be larger than the actors defined! (Actor type = {0})".F(ai.Name));
+
+			if (InitialActorCount < -1)
+				throw new YamlException("InitialActorCount must be -1 or non-negative. Actor type = {0}".F(ai.Name));
+		}
+
+		public override object Create(ActorInitializer init) { return new BaseSpawnerParent(init, this); }
+	}
+
+	public class BaseSpawnerParent : ConditionalTrait<BaseSpawnerParentInfo>, INotifyKilled, INotifyOwnerChanged, INotifyActorDisposing
+	{
+		readonly Actor self;
+		protected readonly BaseSpawnerChildEntry[] ChildEntries;
+
+		IFacing facing;
+		ExitInfo[] exits;
+		RallyPoint rallyPoint;
+
+		protected IReloadModifier[] reloadModifiers;
+
+		public BaseSpawnerParent(ActorInitializer init, BaseSpawnerParentInfo info)
+			: base(info)
+		{
+			self = init.Self;
+
+			// Initialize child entries (doesn't instantiate the children yet)
+			ChildEntries = CreateChildEntries(info);
+
+			for (var i = 0; i < info.Actors.Length; i++)
```

Comment:
Is there a reason we do this in a second loop, and not in `CreateChildEntries` where we already loop anyway? (I don't see one.)

Reply:
It otherwise crashes:

```
Exception of type `System.NullReferenceException`: Object reference not set to an instance of an object
  at OpenRA.TraitDictionary+TraitContainer`1[T].Get (OpenRA.Actor actor) [0x00016] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.TraitDictionary.Get[T] (OpenRA.Actor actor) [0x0000e] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.Actor.Trait[T] () [0x00001] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.Actor..ctor (OpenRA.World world, System.String name, OpenRA.Primitives.TypeDictionary initDict) [0x00166] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.World.CreateActor (System.Boolean addToWorld, System.String name, OpenRA.Primitives.TypeDictionary initDict) [0x00001] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.Mods.RA2.Traits.BaseSpawnerParent.Replenish (OpenRA.Actor self, OpenRA.Mods.RA2.Traits.BaseSpawnerChildEntry entry) [0x00039] in <bbd07fa353dc423792c3b5fce2f06700>:0 
  at OpenRA.Mods.RA2.Traits.CarrierParent.Replenish (OpenRA.Actor self, OpenRA.Mods.RA2.Traits.BaseSpawnerChildEntry entry) [0x00001] in <bbd07fa353dc423792c3b5fce2f06700>:0 
  at OpenRA.Mods.RA2.Traits.BaseSpawnerParent.Replenish (OpenRA.Actor self, OpenRA.Mods.RA2.Traits.BaseSpawnerChildEntry[] childEntries) [0x0002b] in <bbd07fa353dc423792c3b5fce2f06700>:0 
  at OpenRA.Mods.RA2.Traits.CarrierParent.Created (OpenRA.Actor self) [0x00042] in <bbd07fa353dc423792c3b5fce2f06700>:0 
  at OpenRA.Mods.Common.Traits.ConditionalTrait`1[InfoType].OpenRA.Traits.INotifyCreated.Created (OpenRA.Actor self) [0x00001] in <c802dc0798c54c2fac903596f1dc4b3c>:0 
  at OpenRA.Actor.Created () [0x0001e] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.World.CreateActor (System.Boolean addToWorld, System.String name, OpenRA.Primitives.TypeDictionary initDict) [0x0000a] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.World.CreateActor (System.String name, OpenRA.Primitives.TypeDictionary initDict) [0x00001] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.Mods.Common.Traits.Production+<>c__DisplayClass6_0.<DoProduction>b__0 (OpenRA.World w) [0x0000e] in <c802dc0798c54c2fac903596f1dc4b3c>:0 
  at OpenRA.World.Tick () [0x001d9] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x001ed] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.Game.LogicTick () [0x0004a] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.Game.Loop () [0x00108] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.Game.Run () [0x00044] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00013] in <35331ddaa9a142e4b63a72df2041355c>:0 
  at OpenRA.Program.Main (System.String[] args) [0x0004e] in <35331ddaa9a142e4b63a72df2041355c>:0
```

Reply:
This is probably because `CreateChildEntries` is overwritten in the inheriting class. Not very obvious to be honest, but I see now.